tap-version-14-specification.md

```diff
@@ -0,0 +1,659 @@
+and the text after `SKIP` is the explanation.
+
+```tap
+ok 14 - mung the gums # SKIP leave gums unmunged for now
+```
+
+If the SKIP has an explanation, it must be separated from SKIP by a space.
+These tests indicate that a test was not run, or if it was, that its
+success or failure is being temporarily ignored.
+
+Harnesses _must not_ treat failing `SKIP` test points as a test failure.
+
+Harnesses _should_ report `SKIP` test points found as a list of items that
+were not tested, if that is appropriate for their use case.
+
+### YAML Diagnostics
+
+If a Test Point is followed by a 2-space indented block beginning with
+the line `  ---` and ending with the line `  ...`, separated from the Test
+Point only by comments or whitespace, then the block of lines between these
+markers will be interpreted as an inline YAML Diagnostic document.
+
+The YAML encodes a data structure that provides information about the
+preceding Test Point.
+
+For example:
+
+```tap
+not ok 3 - Resolve address
+  ---
+  message: "Failed with error 'hostname peebles.example.com not found'"
+  severity: fail
+  found:
+    hostname: 'peebles.example.com'
+    address: ~
+  wanted:
+    hostname: 'peebles.example.com'
+    address: '85.193.201.85'
+  at:
+    file: test/dns-resolve.c
+    line: 142
+  ...
+```
+
+Currently (March 2022) the data structure represented by a YAML Diagnostic
+block has not been standardized.  TAP14 Harnesses _must_ allow any data
+structures supported by their YAML parser implementation.
+
+A future version of this specification _may_ provide guidance regarding
+YAML Diagnostic fields in common usage.
+
+### Comments
+
+Lines outside of a YAML diagnostic block which begin with a `#` character
+preceeded by zero or more characters of whitespace, are comments.
+
+A Harness _may_ present these to the user, ignore them, or assign meaning
+to certain comments.
+
+A Harness _must not_ treat a test as a failure based on the presence of
+comment lines.  That is, a Harness _must_ ignore any unrecognized comment
+lines.
+
+### Pragmas
+
+A Pragma provides information to a Harness to control its behavior or
+configure it in some way.  Each Pragma line represents a single boolean
+switch which can be set to `true` or `false`.
+
+The structure of a Pragma line is:
+
+- `"pragma "`
+- `+` (true) or `-` (false)
+- key: The name of the field being enabled or disabled.
+
+For example:
+
+```tap
+TAP version 13
+# tell the parser to bail out on any failures from now on
+pragma +bail
+
+# tell the parser to execute in strict mode, treating any invalid TAP
+# as a test failure.
+pragma +strict
+
+# turn off a feature we don't want to be active right now
+pragma -bail
+```
+
+The meaning and availability of keys that may be set by Pragmas are
+implementation-specific.
+
+Harnesses _must not_ treat unrecognized Pragma keys as a test failure.
+However, they _may_ warn if a Pragma is unrecognized, or fail if the
+named flag cannot be set for some reason.
```

Comment:
Node-tap treats `pragma -bail # comment` as a non-TAP line.

Reply:
I wonder what others think. I think I could probably document pragmas with comments above it, but I guess some users would/could also try to put it as a comment/directive next to it? If the spec doesn't say it's invalid and must be treated as non-TAP line, I'd probably add support to it in `tap4j`.

Reply:
If tap4j doesn't already support pragmas at all today, I'd recommend not adding trailing # comments.  Trying to be lazy and limit how much _new_ work we make for ourselves here 😅  We still have child tests to specify.

As written so far in the draft, `pragma -bail # comment` _would_ be under "anything else", because `"bail # comment"` does not match `/[a-zA-Z0-9_-]+/`.

Reply:
d5a477e72d449247fd46062bbead773949c976e2